task/host-sp-comms/src/main.rs

```diff
@@ -391,40 +400,37 @@ impl ServerImpl {
 
             // Receive until there's no more data or we get a 0x00, signifying
             // the end of a corncobs packet.
-            while let Some(byte) = self.uart.try_rx_pop() {
+            'rx: while let Some(byte) = self.uart.try_rx_pop() {
                 ringbuf_entry!(Trace::UartRx(byte));
 
-                if byte == 0x00 {
-                    // Host may send extra cobs terminators; skip any empty
-                    // packets.
-                    if self.rx_buf.is_empty() {
-                        continue;
+                if byte != 0x00 {
+                    // This is the end of a packet; buffer it and continue.
+                    if self.rx_buf.push(byte).is_err() {
+                        // Message overflow - nothing we can do here except
+                        // discard data. We'll drop this byte and wait til we
+                        // see a 0 to respond, at which point our
+                        // deserialization will presumably fail and we'll send
+                        // back an error. Should we record that we overflowed
+                        // here?
                     }
+                    continue 'rx;
+                }
 
-                    // Process message and populate our response into
-                    // `tx_msg_buf`.
-                    let msg_len = self.process_message();
-                    self.rx_buf.clear();
+                // Host may send extra cobs terminators; skip any empty
+                // packets.
+                if self.rx_buf.is_empty() {
+                    continue 'rx;
+                }
 
-                    if let Some(msg_len) = msg_len {
-                        // Encode our outgoing packet.
-                        let pkt_len = corncobs::encode_buf(
-                            &self.tx_msg_buf[..msg_len],
-                            &mut self.tx_pkt_buf[..],
-                        );
-                        self.tx_pkt_to_write = 0..pkt_len;
+                // Process message and set up `self.tx_buf` with our response
+                // (or intermediate state if we don't have a response yet).
+                self.process_message();
+                self.rx_buf.clear();
```

Comment:
Do we ever call `self.process_message()` without then clearing `rx_buf`?  If not, then we may want to put `self.rx_buf.clear()` into `process_message`.

Reply:
This took a bit of refactoring to keep the borrow checker happy, but I think still ended up net positive? 6ee65fb4 if you disagree, let me know!

Reply:
Yeah, looks better to me!